### PR TITLE
Create integration-ReversingLabsTitaniumCloud

### PR DESCRIPTION
**Pull request by secdevopsai**
# ReversingLabs Titanium Cloud Integration

### Overview
ReversingLabs Data provides malware status of the sample. Status can be: malicious, suspicious, known and unknown. Service supports single or bulk queries, with extended option for response data (trust factor, threat level, ...).

### To set up the integration on Demisto:
- Enter credentials
- Enter Base URL

### Use-cases:
File Reputation

#### file
**Input**
file -  Retrieve Malware Presence Status from ReversingLabs
**Context Output**
```
URL.Malicious.Vendor - The vendor that decides whether a URL is malicious.
URL.Malicious.Description - The reason for the vendor decided that a URL is malicious.
DBotScore.Indicator - The indicator that was tested.
DBotScore.Type - The indicator type.
DBotScore.Vendor - Vendor used to calculate the score.
DBotScore.Score - The actual score.
```
**Raw Output**
```
{
    "sample": {
        "analysis": {
            "entries": [
                {
                    "analysis_type": "TC_REPORT",
                    "analysis_version": "3.0.1",
                    "record_time": "2017-11-23T01:42:52",
                    "tc_report": {
                        "info": {
                            "file": {
                                "file_subtype": "Archive",
                                "file_type": "Binary"
                            },
                            "identification": {
                                "name": "ZIP"
                            }
                        }
                    }
                }
            ]
        },
        "md5": "5bafb135e1d7ba0a5acd0fbbeb2a93e1",
        "relationships": {
            "container_sample_sha1": [
                "a1791f9a6e7662a37d41f0e49b2b3ceaa7d140dd"
            ]
        },
        "ripemd160": "5eaada822b132b0fb111a60217ce2a50411fe569",
        "sample_size": 964785,
        "sha1": "a1791f9a6e7662a37d41f0e49b2b3ceaa7d140dd",
        "sha256": "99550ea419a3f49ee1154c9d5ae99b7f925a783028f87286584a0c9fe7bab4f0",
        "sha384": "53a41129de1f3a573d40e240f4f0b4359125082a2071cf0c0528687bee0098a2a83c03eba01110256908b46361f05797",
        "sha512": "fa62217ecfca3137ef8851b55c8ff34c8c1dfac792620335266ceb12f1b63653496e4fd6aa02adf11bc033f11b57559c7b9aa6a09ea337ed42ff0ae8170263e4",
        "sources": {
            "entries": [
                {
                    "record_time": "2017-11-27T10:35:14",
                    "tag": "reversing_labs"
                },
                {
                    "properties": [
                        {
                            "name": "tags",
                            "value": "malware"
                        }
                    ],
                    "record_time": "2017-11-23T01:41:24",
                    "tag": "external_feed"
                }
            ]
        },
        "ssdeep": "24576:mqlpq3zNq1Qm/xP8pWJ7zmJgL3LuQSfE2E2SFAPcA+/wc:mqjgNq1LPAWJ7LLF8Romcn/9"
    }
}
```
**Usage example**
!file file=a1791f9a6e7662a37d41f0e49b2b3ceaa7d140dd

### CODE REVIEW

**Comment by dorsha** [comment]
@anara123

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  be: malicious, suspicious, known and unknown. Service supports single or bulk queries,
+  with extended option for response data (trust factor, threat level, ...).'
+configuration:
+- display: Credentials
+  name: credentials
+  defaultvalue: ""
+  type: 9
+  required: true
+- display: Base URL
+  name: base
```
Comment (anara123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: This param should be mandatory?

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+description: 'ReversingLabs Data provides malware status of the sample. Status can
```
Comment (anara123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: image
The detailed instructions field is missing. Don't we need one, or it is straight forward to add new instance. If yes then there is no need to fill this field.

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        ec['DBotScore'].append({'Indicator': md5, 'Type': 'hash', 'Vendor': 'ReversingLabs', 'Score': dbot_score_map[contents['status']]})
+        ec['DBotScore'].append({'Indicator': sha1, 'Type': 'hash', 'Vendor': 'ReversingLabs', 'Score': dbot_score_map[contents['status']]})
+        ec['DBotScore'].append({'Indicator': sha256, 'Type': 'hash', 'Vendor': 'ReversingLabs', 'Score': dbot_score_map[contents['status']]})
+        return md, ec
+
+
+    if demisto.command() == 'test-module':
+        try:
+            contents = requests.get("https://ticloud-cdn-api.reversinglabs.com/api/databrowser/rldata/query/md5/6a95d3d00267c9fd80bd42122738e726?extended=true&format=json",auth=auth).json()['rl']
```
Comment (anara123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: Is `https://ticloud-cdn-api.reversinglabs.com` is hard coded on purpose? Why not to use the `params.base`?

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+
+    if not demisto.params()['proxy']:
+        del os.environ['HTTP_PROXY']
+        del os.environ['HTTPS_PROXY']
+        del os.environ['http_proxy']
+        del os.environ['https_proxy']
+
+
+    def rldata(hash_value):
```
Comment (anara123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: comment/function description which describe what are the steps happening in the function could help a lot to faster and easily understand the code.

File: TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    taskid: 15bba6b2-1259-4136-8e90-3cacc3bd2554
+    type: regular
+    task:
+      id: 15bba6b2-1259-4136-8e90-3cacc3bd2554
+      version: -1
+      name: Enrich MD5
+      script: ReversingLabs Titanium Core|||file
+      type: regular
+      iscommand: true
+      brand: ReversingLabs Titanium Core
```
Comment (anara123) [inline TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml]: You mean `ReversingLabs Titanium Cloud`? and not `Core`

File: TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml [hunk 6]
```diff
@@ -0,0 +1,216 @@
+id: 01f92c20-349f-4ced-877e-30476d014fe3
+version: 2
+name: ReversingLabsTitaniumCloud-Test
```
Comment (anara123) [inline TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml]: Currently the playbook has no inputs. Incidents needs to have some hashes in order this test playbook enrich the hashes. You need to insert some hard coded hashes into the context/incident and then validate if the enrich tasks really retrieved the correct data. 
The playbook need to be able to run automatically without anyone run it manually or put him some data

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                'regex':r'([a-fA-F\d]{64})'
+            }
+        }
+
+        if len(hash_value) not in type_dict.keys():
+            demisto.results({ 'Type' : entryTypes['error'], 'ContentsFormat' : formats['text'], 'Contents' : "Not Valid Hash"})
+            sys.exit(0)
+
+        if not re.match(type_dict[len(hash_value)]['regex'],hash_value):
+            demisto.results({ 'Type' : entryTypes['error'], 'ContentsFormat' : formats['text'], 'Contents' : "Not Valid Hash1" })
```
Comment (orius123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: `Hash1` is a bit confusing. maybe not valid by regex, has illegal chars?

File: Integrations/integration-ReversingLabsTitaniumCloud.yml [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+    if demisto.command() == 'file':
+        hash_value = demisto.args()['file']
+        rldata(hash_value)
+  type: python
+  commands:
+  - name: file
+    arguments:
+    - name: file
+      required: true
```
Comment (orius123) [inline Integrations/integration-ReversingLabsTitaniumCloud.yml]: should also be default

File: Integrations/integration-ReversingLabsTitaniumCloud.yml (hunk 3 above)
Comment (secdevopsai) [reply Integrations/integration-ReversingLabsTitaniumCloud.yml]: Good catch

File: TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml (hunk 6 above)
Comment (secdevopsai) [reply TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml]: I'll add manual input

File: TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml (hunk 5 above)
Comment (secdevopsai) [reply TestPlaybooks/playbook-ReversingLabsTitaniumCloud-Test.yml]: Yes

File: Integrations/integration-ReversingLabsTitaniumCloud.yml (hunk 7 above)
Comment (secdevopsai) [reply Integrations/integration-ReversingLabsTitaniumCloud.yml]: Typo

File: Integrations/integration-ReversingLabsTitaniumCloud.yml (hunk 8 above)
Comment (secdevopsai) [reply Integrations/integration-ReversingLabsTitaniumCloud.yml]: Correct

**Comment by secdevopsai** [comment]
Further revisions will be needed

**Pull request closed without merging**